doc/syslog/syslog-rate-limit-design.md

```diff
@@ -0,0 +1,376 @@
+{% endif %}
+```
+
+#### rsyslog-container.conf.j2
+
+```
+{% set rate_limit_interval = '300' %}
+{% set rate_limit_burst = '20000' %}
+
+{% if SYSLOG_CONFIG_FEATURE is defined %}
+{% if container_name in SYSLOG_CONFIG_FEATURE %}
+{% if 'rate_limit_interval' in SYSLOG_CONFIG_FEATURE[container_name]%}
+{% set rate_limit_interval = SYSLOG_CONFIG_FEATURE[container_name]['rate_limit_interval'] %}
+{% endif %}
+{% if 'rate_limit_burst' in SYSLOG_CONFIG_FEATURE[container_name]%}
+{% set rate_limit_burst = SYSLOG_CONFIG_FEATURE[container_name]['rate_limit_burst'] %}
+{% endif %}
+{% endif %}
+{% endif %}
+
+$SystemLogRateLimitInterval {{ rate_limit_interval }}
+$SystemLogRateLimitBurst {{ rate_limit_burst }}
+```
+
+#### docker_image_ctl.j2
+
+Function `updateSyslogConf` in the template currently only works for multi ASIC platform, it shall be extended to work for both multi ASIC and single ASIC platform. The only difference between multi ASIC and single ASIC platform is the target IP of rsyslog.
+
+- Mulit ASIC: target IP of rsyslog is get from docker0 IP which is already implemented in docker_image_ctl.j2
+- Single ASIC: target IP of rsyslog is always 127.0.0.1.
+
+Meanwhile, the file https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-base/etc/rsyslog.conf as well as any code processing it shall be removed.
+
+#### APP extension
+
+APP extension shall be able to expose its constant syslog capability by adding a new field to manifest:
+
+```
+root@sonic:/home/admin# spm show package manifest what-just-happened
+{
+    ...
+    "service": {
+        ...
+        "syslog": {
+            "support-rate-limit": "true",
+            "customized-rate-limit": "false"
+        }
+        ...
+    }
+    ...
+}
+```
+
+support-rate-limit: indicates if this APP extension supports configuring rate limit. This field affected CLI behavior: if true, CLI shall put rate limit configuration to CONFIG DB; otherwise, CLI shall reject the configuration.
+customized-rate-limit: indicates if this APP extension is using "SONiC official way" to configure rate limit. Effective only if support-rate-limit is true. This field affected CLI behavior: if false, CLI shall generated rsyslogd configuration file by rendering rsyslog-container.conf.j2, coping it to relevant container and restart rsyslogd by supervisorctl command; otherwise, app extension shall handle CONFIG DB change and set log rate limit accordingly. "SONiC official way" means:
+
+- Using rsyslogd
+- Configuration of rsyslogd is generated by rsyslog-container.conf.j2
+- rsyslogd can be managed by supervisorctl
+
+
+
+The capability shall also be saved to CONFIG DB to allow easy access by management plane:
+
+```
+root@sonic:/home/admin# cat /etc/sonic/config_db.json
+{
+    "FEATURE": {
+        ...
+        "what-just-happened": {
+            ...
+            "support_syslog_rate_limit": "True",
+            "has_customized_syslog_rate_limit": "False"
```

Comment:
Can we make it configurable for rsyslog to suppress duplicate messages?

Reply:
It is out of the feature scope. We may need a new feature on top of this feature to support it.